### ISSUE TITLE
Warnings Are Not Raised For Code In Conditional Compilation That Is False At Compile Time

### ISSUE THREAD

**Issue by martincostello** [DotNetAnalyzers/StyleCopAnalyzers#1750]
Consider the following piece of code:

```
#if TFS_BUILD
        [Ignore]              // Do not run as part of CI
#endif
        [TestMethod]
        [TestCategory("LocalOnly")]
        public void Test_Something()
        {
            // Test code
        }
```
When compiled locally, the code has no warnings. When compiled with ```TFS_BUILD``` defined, it raises SA1025.

This leaves CI builds open to breakage as warnings can be introduced by changes that can't be found easily locally, which then go on to break a build. I imagine similar issues could be found through code that uses ```#if DEBUG```, ```#if TRACE``` etc.

I can see that given that StyleCop is now a compile-time analyser that it would take the compilation settings into account during the analysis, however given that I'd imagine that in most cases people would want code analysis done on the file _as a whole_, this behaviour is sub-optimal.

**Comment by sharwell** [comment]
Code inside preprocessor directives which are disabled is reported as excluded code by Roslyn. It's not possible to perform any meaningful analysis inside these blocks unless you (for this case) define `TFS_BUILD`. I'm marking this as **wontfix**, but in reality it's more of a "can't fix".

FWIW, the correct way to exclude these test methods for a TFS build is to pass the following flag to **vstest.console**:

```
/TestCaseFilter:"TestCategory!=LocalOnly"
```

**Comment by martincostello** [comment]
@sharwell Fair enough if it can't be fixed. We also use the test case filters as well, this is just an additional thing we put it for some other unrelated reasons.

**Comment by sharwell** [comment]
:memo: If you encounter a case where analysis is incorrect when using `#if true`, please do file a bug. I'm having trouble finding an example right now, but we've fixed a few over time.
